Add support for generating multiple bls keys for a given shard and node

SebastianJ opened the pull request:
This PR adds support for generating multiple bls keys targeting a specific shard for a given node/network.

E.g:
```
$ ./hmy keys generate-bls-keys --node https://api.s0.os.hmny.io --shard 0 --count 3
{
  "public-key": "ca5fb83dd63455f6ce0f06960e05ff3ab2b3deffa6bad0530c2fc684b00ee55833f189501f97aa9202153bf38efa2b14",
  "private-key": "090408670a8e40a32ac73ea11cbe680577f462576ea459538f3b877b4d809e19",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/ca5fb83dd63455f6ce0f06960e05ff3ab2b3deffa6bad0530c2fc684b00ee55833f189501f97aa9202153bf38efa2b14.key"
}
{
  "public-key": "30f3963cfa95b8f774691c42cf2560b4ddc49c5e80e97350d53f31bff8e3544d43971c2487f8c24a51e7c99a8da35384",
  "private-key": "cbb6a255d4e5433f1764c03e171a1803d720eb0bc348f72ef32c2b41af37ce2f",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/30f3963cfa95b8f774691c42cf2560b4ddc49c5e80e97350d53f31bff8e3544d43971c2487f8c24a51e7c99a8da35384.key"
}
{
  "public-key": "a722dad939dc15d8555dbdc94c954e497c112162246e0b7fe3e53f12121e5852dcc3c286f263e5f37a607965f95af218",
  "private-key": "280472f28078b36175c7cb1e3a54e47467cfe957606dc0230d9dd6d681ad294b",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/a722dad939dc15d8555dbdc94c954e497c112162246e0b7fe3e53f12121e5852dcc3c286f263e5f37a607965f95af218.key"
}
```

Verification:
```
$ ./hmy utility shard-for-bls --node https://api.s0.os.hmny.io ca5fb83dd63455f6ce0f06960e05ff3ab2b3deffa6bad0530c2fc684b00ee55833f189501f97aa9202153bf38efa2b14
{"shard-id":0}
```

```
$ ./hmy utility shard-for-bls --node https://api.s0.os.hmny.io 30f3963cfa95b8f774691c42cf2560b4ddc49c5e80e97350d53f31bff8e3544d43971c2487f8c24a51e7c99a8da35384
{"shard-id":0}
```

```
$ ./hmy utility shard-for-bls --node https://api.s0.os.hmny.io a722dad939dc15d8555dbdc94c954e497c112162246e0b7fe3e53f12121e5852dcc3c286f263e5f37a607965f95af218
{"shard-id":0}
```

gupadhyaya commented:
@SebastianJ can you repeat the tests for some other network as well and post the results?
Also, does this need same passphrase for all the bls keys? Can you extend it to support different passphrase?

gupadhyaya commented on pkg/keys/bls.go:
```diff
@@ -30,25 +32,40 @@ func GenBlsKeys(passphrase, filePath string) error {
+	shardCount := len(shardingStructure)
+	generatedCount := uint32(0)
+
+	for {
```
can you use traditional for loop?

SebastianJ commented:
Ok, so I added support for using different passwords for each bls key:

```
$ ./hmy keys generate-bls-keys --node https://api.s0.os.hmny.io --shard 0 --count 3 --passphrase
Enter passphrase:
Repeat the passphrase:

Enter passphrase:
Repeat the passphrase:

Enter passphrase:
Repeat the passphrase:

{
  "public-key": "fa885e9b07d1ec412e6547defbbaa0596c6218534733c0fd54f1427843bd6e4af81710515013981c9bd7bda82e4a6d04",
  "private-key": "1440edf25bd70c4436acf5786a1e8677546fe65650f37792d4dcf1c8fd39df6a",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/fa885e9b07d1ec412e6547defbbaa0596c6218534733c0fd54f1427843bd6e4af81710515013981c9bd7bda82e4a6d04.key"
}
{
  "public-key": "5044d4a8727aad0bfbc12101977a9b6fef1c7c508b5eb1514b3b57552b38e44a80d39adbff2a9757c04d55755157ba90",
  "private-key": "4b7e4019807e85edbaac866f99cf4305bcde975b3ea660c61eef6aeedfcdf528",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/5044d4a8727aad0bfbc12101977a9b6fef1c7c508b5eb1514b3b57552b38e44a80d39adbff2a9757c04d55755157ba90.key"
}
{
  "public-key": "dc4d47065e9dfdcbdef6874a7a2bc5db92e7e708c150e830afe0b4594f98029d9b8f43f40b8bef254595b75435dd2f18",
  "private-key": "0cd62db911a8bb2fc15c13e813c57e6fce902c1a16b334d85c561daeda0e730a",
  "encrypted-private-key-path": "$GOPATH/src/github.com/SebastianJ/go-sdk/dc4d47065e9dfdcbdef6874a7a2bc5db92e7e708c150e830afe0b4594f98029d9b8f43f40b8bef254595b75435dd2f18.key"
}
```

I also added unit tests for regular bls key generation + multi bls key generation using the node endpoints:

- https://api.s0.os.hmny.io
- https://api.s0.ps.hmny.io
- https://api.s0.stn.hmny.io
- https://api.s0.b.hmny.io
- https://api.s0.t.hmny.io

To run the tests:

```
$ cd pkg/keys && go test ./... && cd ../..
ok  	github.com/harmony-one/go-sdk/pkg/keys	7.756s
```

I also refactored some of the bls key generation functions a bit to make it easier to implement unit tests.

denniswon commented on pkg/keys/bls.go:
```diff
@@ -11,44 +11,65 @@ import (
+}
+
+// GenMultiBlsKeys - generate multiple BLS keys for a given shard and node/network
+func GenMultiBlsKeys(passphrases []string, filePath string, node string, count uint32, shardID uint32) error {
```
any validation conditions to make sense for count? :man_shrugging:

SebastianJ replied:
Should there be? If they generate too many bls keys for a given network they can't start the node anyways, so the node will validate/set a hard limit anyways. Or are you referring to something else?

denniswon replied:
that was what I was referring to, the limit imposed on running the node. Yeah, if it is imposed when running the node, it seems duplicative.

SebastianJ commented on cmd/subcommands/keys.go:
```diff
@@ -321,6 +323,29 @@ func keysSub() []*cobra.Command {
+		"absolute path of where to save encrypted bls private keys")
+	cmdGenerateMultiBlsKeys.Flags().BoolVar(&userProvidesPassphrase, "passphrase", false, ppPrompt)
+	cmdGenerateMultiBlsKeys.Flags().StringVar(&passphraseFilePath, "passphrase-file", "", "path to a file containing the passphrase")
+	cmdGenerateMultiBlsKeys.Flags().Uint32Var(&blsShardID, "shard", 0, "which shard to create bls keys for")
+	cmdGenerateMultiBlsKeys.Flags().Uint32Var(&blsCount, "count", 1, "how many bls keys to generate")
+
```
The purpose of this specific functionality is to only generate bls keys for a given shard - e.g. to be used for starting a multi-bls node (which can only be started using keys belonging to the same shard). The regular bls key function still randomly generates bls keys without targeting a specific shard and can be used to generate random bls keys.

denniswon replied:
okay. Would be also nice to add such info to the cmd description.

SebastianJ commented on pkg/keys/bls.go:
```diff
@@ -11,44 +11,65 @@ import (
-		return err
+
+	for _, blsKey := range blsKeys {
+		if err = writeBlsKeyToFile(filePath, blsKey); err != nil {
```
You're correct - I missed this. If you don't pass an explicit `filePath` it'll work since every key will be generated based on the key's publicHexKey, so keys won't get overwritten. I'll switch it to a map of path => pw so that custom paths are also supported.

SebastianJ commented:
> Could we get the output to be valid JSON? Could just make it a JSON array.

Yup, will fix it

SebastianJ commented:
Ok, added support for specifying unique file path and password combinations for each key. Whenever something is passed via `--bls-file-path`, e.g. `--bls-file-path stdin` the user will be prompted to enter an absolute path for each key they want to generate.

Given that the user will be prompted for a password for each key I thought it was better to do it this way versus e.g. accepting a comma delimited list of paths supplied to `--bls-file-path`.

Here's an example generating 3 keys for shard 0 (for OSTN/or a similar network with the same set of shards), using custom passwords and file paths for each key:

```
$ ./hmy keys generate-bls-keys --node https://api.s0.os.hmny.io --shard 0 --count 3 --passphrase --bls-file-path stdin
Enter absolute path for key #1:
/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/1.key
Enter passphrase:
Repeat the passphrase:

Enter absolute path for key #2:
/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/2.key
Enter passphrase:
Repeat the passphrase:

Enter absolute path for key #3:
/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/3.key
Enter passphrase:
Repeat the passphrase:

[
  {
    "public-key": "15b47ec6ac349854db1023f4a97812167065cd38dc747fded76b639babf4264f98215b3041f6efb13f658c9c6890778c",
    "private-key": "9c037741b08ccd9ee6f93cc5f24049a78446299a1bb55c35cfcd98f3696cc947",
    "encrypted-private-key-path": "/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/1.key"
  },
  {
    "public-key": "184bf1a5268c8d31f66600dbbcb606efcff3f857fb1cb8089f523191e3f18aaa997df18bf2ebb098318bceffe9fb8b10",
    "private-key": "4af74ac02156a02bc90f35d9f87b1a1347d9e339937623710b2e119143aaa036",
    "encrypted-private-key-path": "/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/2.key"
  },
  {
    "public-key": "10faba1cdde9ea74732b5e08b606483be8e5da9b648b4d11e98998cf5e549524eb0ee00e415fd6cb3b85d441f04e2314",
    "private-key": "253c78052b533a7e136801e2ef62cea3e7adb8a63ae9270920b7bad57d745806",
    "encrypted-private-key-path": "/Users/sebastianjohnsson/Programming/Golang/src/github.com/SebastianJ/go-sdk/3.key"
  }
]
```

Verification that the keys do in fact belong to shard 0:
```
$ ./hmy --node https://api.s0.os.hmny.io utility shard-for-bls 15b47ec6ac349854db1023f4a97812167065cd38dc747fded76b639babf4264f98215b3041f6efb13f658c9c6890778c
{"shard-id":0}

$ ./hmy --node https://api.s0.os.hmny.io utility shard-for-bls 184bf1a5268c8d31f66600dbbcb606efcff3f857fb1cb8089f523191e3f18aaa997df18bf2ebb098318bceffe9fb8b10
{"shard-id":0}

$ ./hmy --node https://api.s0.os.hmny.io utility shard-for-bls 10faba1cdde9ea74732b5e08b606483be8e5da9b648b4d11e98998cf5e549524eb0ee00e415fd6cb3b85d441f04e2314
{"shard-id":0}
```

Run the updated unit tests:
```
$ cd pkg/keys && go test ./... && cd ../..
ok  	github.com/harmony-one/go-sdk/pkg/keys	6.854s
```

denniswon commented:
LGTM

denniswon commented:
@SebastianJ @Daniel-VDM 
This PR breaks the current build for go-sdk repo.
```
dennis.won@Jongs-MacBook-Pro:~/harmony-one/go-sdk (master) $ make
source /Users/dennis.won/go/src/github.com/harmony-one/harmony/scripts/setup_bls_build_flags.sh && GO111MODULE=on go build -o ./dist/hmy -ldflags="-X main.version=v359 -X main.commit=833256e-dirty -X main.builtAt=2020-04-02T05:40:36-0700 -X main.builtBy=dennis.won@harmony.one" cmd/main.go
# github.com/harmony-one/go-sdk/pkg/keys
pkg/keys/bls.go:38:18: undefined: shard.BlsPublicKey
note: module requires Go 1.13.6
make: *** [all] Error 2
dennis.won@Jongs-MacBook-Pro:~/harmony-one/go-sdk (master) $ go version
go version go1.13.6 darwin/amd64
```

SebastianJ commented:
The reason was that #199 was merged before this PR finally got accepted, and this PR seems to have reverted the Bls -> BLS change #199 fixed. Seems Bls -> BLS was fixed in #202 again.
